scene/audio/audio_stream_player_internal.cpp

```diff
@@ -0,0 +1,253 @@
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "audio_stream_player_internal.h"
+
+#include "scene/main/node.h"
+#include "scene/scene_string_names.h"
+#include "servers/audio/audio_stream.h"
+
+void AudioStreamPlayerInternal::_set_process(bool p_enabled) {
+	if (physical) {
+		node->set_physics_process_internal(p_enabled);
+	} else {
+		node->set_process(p_enabled);
+	}
+}
+
+void AudioStreamPlayerInternal::process() {
+	Vector<Ref<AudioStreamPlayback>> playbacks_to_remove;
+	for (Ref<AudioStreamPlayback> &playback : stream_playbacks) {
+		if (playback.is_valid() && !AudioServer::get_singleton()->is_playback_active(playback) && !AudioServer::get_singleton()->is_playback_paused(playback)) {
+			playbacks_to_remove.push_back(playback);
+		}
+	}
+	// Now go through and remove playbacks that have finished. Removing elements from a Vector in a range based for is asking for trouble.
+	for (Ref<AudioStreamPlayback> &playback : playbacks_to_remove) {
+		stream_playbacks.erase(playback);
+	}
+	if (!playbacks_to_remove.is_empty() && stream_playbacks.is_empty()) {
+		// This node is no longer actively playing audio.
+		active.clear();
+		_set_process(false);
+	}
+	if (!playbacks_to_remove.is_empty()) {
+		node->emit_signal(SNAME("finished"));
+	}
+}
+
+void AudioStreamPlayerInternal::ensure_playback_limit() {
+	while (stream_playbacks.size() > max_polyphony) {
+		AudioServer::get_singleton()->stop_playback_stream(stream_playbacks[0]);
+		stream_playbacks.remove_at(0);
+	}
+}
+
+void AudioStreamPlayerInternal::notification(int p_what) {
+	switch (p_what) {
+		case Node::NOTIFICATION_ENTER_TREE: {
+			if (autoplay && !Engine::get_singleton()->is_editor_hint()) {
+				play_callable.call(0.0);
+			}
+			set_stream_paused(!node->can_process());
+		} break;
+
+		case Node::NOTIFICATION_EXIT_TREE: {
+			set_stream_paused(true);
+		} break;
+
+		case Node::NOTIFICATION_INTERNAL_PROCESS: {
+			process();
+		} break;
+
+		case Node::NOTIFICATION_PREDELETE: {
+			for (Ref<AudioStreamPlayback> &playback : stream_playbacks) {
+				AudioServer::get_singleton()->stop_playback_stream(playback);
+			}
+			stream_playbacks.clear();
```

Comment:
This is code from original AudioStreamPlayer, which did only this. I'd call `stop()` if I wanted to do more, but unified it into this part. The flag is irrelevant at that point.

tbh this made me realize that `setsetplay` is irrelevant too.

Reply:
> This is code from original AudioStreamPlayer, which did only this

True. But AudioStreamPlayer2D & *3D cleared `active`, because they called `stop()`.

Whether the flag is relevant or not. No idea. At least, afaik, the audio system does not use that.

Similar with `setplay`.

I don't know how `NOTIFICATION_PREDELETE`, `AudioStreamPlayerInternal::_set_process` and `AudioStreamPlayerInternal::process` exactly interact with each other. The order of execution. And what happens if the 2d/3d audio player node is called with `play()` and then immediately deleted. Or similar edge cases.

The feedback regarding `active` and `setplay` is about preserving the original behaviour - and make it consistent among the audio players.

No idea if they were a "just in case" measure or if they actively prevented edge cases. So, if you know they don't matter, then it's completely fine with me to not reset these variables.

Reply:
`NOTIFICATION_PREDELETE` is received just before the object is freed. The object does not call any method anymore (except destructor), even deferred calls.

Reply:
Ah, then as long as `NOTIFICATION_INTERNAL_PHYSICS_PROCESS` is not executed afterwards, then it sounds safe. Thanks!